src/treelearner/feature_histogram.hpp

```diff
@@ -609,29 +674,65 @@ class FeatureHistogram {
     double sum_left_gradient = grad;
     double sum_right_gradient = sum_gradient - sum_left_gradient;
     // current split gain
-    double current_gain =
-        GetLeafGain<true, true>(sum_right_gradient, sum_right_hessian,
-                                meta_->config->lambda_l1, l2,
-                                meta_->config->max_delta_step) +
-        GetLeafGain<true, true>(sum_left_gradient, sum_left_hessian,
-                                meta_->config->lambda_l1, l2,
-                                meta_->config->max_delta_step);
+    double current_gain;
+    bool use_smoothing = meta_->config->path_smooth > kEpsilon;
+    if (use_smoothing) {
```

Comment:
for these `GatherInfoxxx` functions, maybe you can use `GetLeafGain<true, true, true>`, and check `path_smooth > kEpsilon' inside?

Reply:
See comment above.

Reply:
okay, maybe we can make `GatherInfoxxx` to template function too.  For example, 'template<bool USE_SMOOTHING> GatherInfoForThresholdNumericalInner'. And check the use_smoothing in GatherInfoForThresholdNumerical. By this way, we can reduce these if ... else.

Reply:
Changed as suggested.